extensions/mssql/src/credentialstore/sqlCredentialService.ts

```diff
@@ -0,0 +1,144 @@
+
+	private static readonly messagesTypes: RPCMessageType[] = [
+		Contracts.DeleteCredentialRequest.type,
+		Contracts.SaveCredentialRequest.type,
+		Contracts.ReadCredentialRequest.type
+	];
+
+	public static asFeature(context: AppContext): ISqlOpsFeature {
+		return class extends SqlCredentialService {
+			private _secretStorage: SecretStorage;
+			private _useNativeCredentialService: boolean;
+			private _passwordsMigrated: boolean;
+
+			constructor(client: SqlOpsDataClient) {
+				super(context, client);
+				this._secretStorage = context.extensionContext.secrets;
+				this._passwordsMigrated = context.extensionContext.globalState.get(Utils.configPasswordsMigrated);
+				this._useNativeCredentialService = Utils.useNativeCredentialsEnabled();
+			}
+
+			override fillClientCapabilities(capabilities: ClientCapabilities): void {
+				Utils.ensure(Utils.ensure(capabilities, 'credentials')!, 'credentials')!.dynamicRegistration = true;
+			}
+
+			override async initialize(capabilities: ServerCapabilities): Promise<void> {
+				this.register(this.messages, {
+					id: UUID.generateUuid(),
+					registerOptions: undefined
+				});
+				if (this._useNativeCredentialService && !this._passwordsMigrated) {
+					await this.migratePasswords();
+					this._passwordsMigrated = true;
+					await context.extensionContext.globalState.update(Utils.configPasswordsMigrated, this._passwordsMigrated);
+				}
+			}
+
+			/**
+			 * Removes a credential for a given connection profile from sqltoolsservice
+			 * and adds it to secret storage
+			 */
+			private async migrateCredential(conn: azdata.connection.ConnectionProfile): Promise<boolean> {
+				const credentialId = Utils.formatCredentialId(conn);
+				// read password from every saved password connection
+				const credential = await this._client.sendRequest(Contracts.ReadCredentialRequest.type, { credentialId, password: undefined });
+				if (credential.password) {
+					const password = credential.password;
+					// save it in secret store
+					await this._secretStorage.store(credentialId, password);
+					// check if it's saved
+					const savedPassword = await this._secretStorage.get(credentialId);
+					if (savedPassword === password) {
+						// delete from tools service
+						const result = await this._client.sendRequest(Contracts.DeleteCredentialRequest.type,
+							{ credentialId, password: undefined });
+						return result;
+					} else {
+						return false;
+					}
+				} else {
+					return false;
+				}
+			}
+
+			/**
+			 *
+			 * @returns Migrates all saved credentials to the native credential system
+			 */
+			private async migratePasswords(): Promise<void> {
+				const connections = await azdata.connection.getConnections(false);
+				const savedPasswordConnections = connections.filter(conn => conn.savePassword === true);
+				for (let i = 0; i < savedPasswordConnections.length; i++) {
+					let conn = savedPasswordConnections[i];
+					await this.migrateCredential(conn);
```

Comment:
I would suggest surrounding this in a try/catch so that if an error occurs migrating one credential it doesn't block the others from being migrated. Although in that case we should make sure we handle that correctly - what's the behavior we want if we can't migrate some/all of the credentials?